File: USACO/2006-2007/Feb 2007/Silver/sparty.java
// "Silver Cow Party"

import java.util.*;
import java.io.*;

public class sparty {
	public static void main(String[] args) throws IOException {
		String file = "";
		// BufferedReader br = new BufferedReader(new FileReader(new File(file + ".in")));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int X = Integer.parseInt(st.nextToken());
		
		ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
		ArrayList<ArrayList<Pair>> adjR = new ArrayList<>();
		
		for(int i = 0; i < N; i++) {
			adj.add(new ArrayList<>());
			adjR.add(new ArrayList<>());
		}
		
		for(int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken())-1;
			int b = Integer.parseInt(st.nextToken())-1;
			int t = Integer.parseInt(st.nextToken());
			adj.get(a).add(new Pair(b, t));
			adjR.get(b).add(new Pair(a, t));
		}
		
		br.close();
		
		int result = 0;
		int[] dist1 = new int[N], dist2 = new int[N];
		Arrays.fill(dist1, 2000000000);
		Arrays.fill(dist2, 2000000000);
		
		dijkstra(adj, dist1, X);
		dijkstra(adjR, dist2, X);
		//PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(new File(file + ".out"))));
		
		for(int i = 0; i < N; i++) result = Math.max(result,  dist1[i] + dist2[i]);
		System.out.println(result);
		//out.println(result);
		//out.close();
	}
	
	
	public static void dijkstra(ArrayList<ArrayList<Pair>> adj, int[] dist, int X) {
		int N = dist.length;
		boolean[] mark = new boolean[N];
		PriorityQueue<Pair> toVisit = new PriorityQueue<>();
		toVisit.add(new Pair(X-1, 0));
		
		while(!toVisit.isEmpty()) {
			Pair curr = toVisit.remove();
			int id = curr.to, d = curr.weight;
			if(mark[id]) continue;
			mark[id] = true;
			
			for(Pair p: adj.get(id)) {
				int to = p.to, weight = p.weight;
				if(!mark[to] && d + weight < dist[to]) {
					dist[to] = weight + d;
					toVisit.add(new Pair(to, weight + d));
				}
			}
		}
		
	}
	
	public static class Pair implements Comparable<Pair>{
		int to, weight;
		
		Pair(int to, int weight) {
			this.to = to;
			this.weight = weight;
		}
		
		public int compareTo(Pair other) {
			return weight - other.weight;
		}
	}
	
}

File: USACO/2006-2007/Feb 2007/Silver/sparty.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#define MAXN 1005

using namespace std;

typedef pair<int, int> pii;

int N, M, X;
int dist1[MAXN], dist2[MAXN];
vector<pii> adj[MAXN], reversed[MAXN];

void dijkstra(int dist[], vector<pii> adj[]);

int main() {
	cin >> N >> M >> X;
	for(int i = 0; i < N; i++) {
		dist1[i] = 2000000000;
		dist2[i] = 2000000000;
	}
	
	for(int i = 0; i < M; i++) {
		int a, b, t;
		cin >> a >> b >> t;
		a--; b--;
		adj[a].push_back(pii(t, b));
		reversed[b].push_back(pii(t, a));
	}

	dijkstra(dist1, adj);
	dijkstra(dist2, reversed);	
	int result = 0;
	for(int i = 0; i < N; i++) {
		result = max(result, dist1[i] + dist2[i]);
	}
	
	cout << result << endl;
}

void dijkstra (int dist[], vector<pii> adj[]) {
	int N = sizeof(dist);
	dist[X-1] = 0;
	priority_queue<pii, vector<pii>, greater<pii>> toVisit;
	toVisit.push(pii(0, X-1));
	while(!toVisit.empty()) {
		pii curr = toVisit.top();
		toVisit.pop();
		int id = curr.second, d = curr.first;
		if(dist[id] != d) continue;
		
		for(pii p: adj[id]) {
			if(d + p.first < dist[p.second]) {
				dist[p.second] = d + p.first;
				toVisit.push(pii(d + p.first, p.second));
			}
		}
	}
}
